### ISSUE TITLE
Missing newton_wrap.cs class

### ISSUE THREAD

**Issue by FerryIcker** [svenberra/NewtonUnityPlugin#2]
Hello Sven,
I can't compile the plugin. in Visual Studio 2015. The File "newton_wrap.cs" is missing.
Can you help?

Best regards Ferry

**Comment by svenberra** [comment]
Make sure you build the whole solution and not only a specific project. The C++ project NewtonWrapper will build first and trigger SWIG to generate the file newton_wrap.cs which in turn is used by the C# project NewtonPlugin.

I updated the readme because recently the Newton SDK rearranged some folders so the Newton SDK project that you need to build before this plugin is now under SDK\Projects instead of Packages\Projects.

Regards, Sven-Bertil

**Comment by FerryIcker** [comment]
Hello,
thanks for fast answer. 
I solved it.
 1.) I moved the hole project in a shorter Path without spaces like c:\project\NGD\..
 2.) changing the Enviromentvariable NEWTON_DYNAMICS
 3.) compiled the newton-dynamics-master
3.)  compiled the Newtonwrapper in the Project NewtonPlugin separetly.
4.) and then the whole Project.
Many thanks

**Comment by FerryIcker** [comment]
Hello again,
Sorry, but I dont know who can help.

Is the callbackfunktion for "NewtonMaterialSetCollisionCallback" in your
newtonunityplugin implemented?

It seems to by not. I am right?
Best regards
Ferry




<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
Virenfrei.
www.avast.com
<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
<#DAB4FAD8-2DD7-40BB-A1B8-4E2AA1F9FDF2>

2017-11-16 11:59 GMT+01:00 svenberra <notifications@github.com>:

**Comment by svenberra** [comment]
Hi.

We intended to use that callback to handle contacts at first but Unity wouldn't play nice with c++ callbacks and multithreading so we fell back to iterating through the bodies ContactJoints and trigger the OnCollision method in NewtonBodyScript instead. Unfortunately for now all you can get is the call to OnCollision with info which body was collided with. But not which collider was involved, material or collisioninfo such as collision normal, force etc.
We will probably have another go at this and implement something more useful but for now only a simple collision notification is implemented.

Regards, Sven-Bertil

22 november 2017 00:25:02 +01:00, skrev FerryIcker <notifications@github.com>:

**Comment by FerryIcker** [comment]
Hi,many thanks for your fast answer,

I hope you will found a solution. It would be greate.

Here is a stuff what for i need it.
This example I wrote in Delphi about 5 years ago.

It is a Sorter controlled by siemens SPS S7

https://youtu.be/2Ukrt1reX_k

best regards

2017-11-22 8:57 GMT+01:00 svenberra <notifications@github.com>:

**Comment by svenberra** [comment]
Nice simulation.
What is it you need from the callback? Collision normals, force?

**Comment by FerryIcker** [comment]
Hi sven,

May be you speek German? i am.

I put the callback funktion Code here.

best regards Ferry


procedure TConveyor.CallBackConv(const ccontact: PNewtonJoint; timestep:
Single; threadIndex: Integer);
var
  Material: PNewtonMaterial;
  ThisContact: PNewtonJoint;
  ActualSpeed, NewAcceleration, MinAcc, MaxAcc: Single;
  BeltDirection: TVector;
  diff: Single;
  rotMatrix: TMatrix;

  function EnsureRange(NewAcceleration, MinAcc, MaxAcc: Single): Single;
  begin
    Result := 0;
    if (NewAcceleration > MinAcc) and (NewAcceleration < MaxAcc) then
      Result := NewAcceleration
    else if NewAcceleration >= MaxAcc then
      Result := MaxAcc
    else if NewAcceleration <= MinAcc then
      Result := MinAcc;
  end;

begin
  MinAcc := -1000;
  MaxAcc := 1000;

  if (abs(FBandV_Ist) > 0.001) then
  begin
    ThisContact := NewtonContactJointGetFirstContact(ccontact);
    // Anpassung Rotation an Newtonsystem
    diff := DegToRad(90);
    rotMatrix := CreateRotationMatrix(Up.AsVector, diff);
    BeltDirection := VectorTransform(Self.AbsoluteDirection, rotMatrix);

    // BeltDirection := Self.AbsoluteDirection;
    while ThisContact <> nil do
    begin
      Material := NewtonContactGetMaterial(ThisContact);
      NewtonMaterialContactRotateTangentDirections(Material,
@BeltDirection);
      ActualSpeed := NewtonMaterialGetContactTangentSpeed(Material, 0);
      NewAcceleration := (FBandV_Ist - ActualSpeed) / timestep;
      NewAcceleration := EnsureRange(NewAcceleration, MinAcc, MaxAcc);
      NewtonMaterialSetContactElasticity(Material, 0);
      NewtonMaterialSetContactTangentAcceleration(Material,
NewAcceleration, 0);
      ThisContact := NewtonContactJointGetNextContact(ccontact,
ThisContact);
    end;
  end;
end;


2017-11-23 9:52 GMT+01:00 svenberra <notifications@github.com>:

**Comment by svenberra** [comment]
Hi.

Unfortunately I don't speak German, I'm swedish.
Correction, I do speak some German but only at tourist level.
"Mein name ist Sven und ich lebe in Schweden"
"Wo is das nächste hotel?"
"Rechts, links oder geradeaus?"

That's about the limits of what I know. :)

Regarding contacts, reading your code i assume you manipulate the contacts to make the boxes move on the convyer belt.

As I mentioned earlier we couldn't get the contact callback to work with Unity and instead iterate the contacts post update. I will have to check with Julio if the contacts can be manipulated post update. If the contact functions only work when called from inside a callback we have to find a way to deal with that. I suspect the callback is triggered between a collision and integration phase and therefore manipulating the contacts after the integration phase won't work. But I will check with Julio first.

There is another way to "simulate" conveyor belts though. If you have a static or kinematic body and set it's velocity, then bodies resting on it should slide like it was a convyor belt. I guess it would be trickier with your spinning convyerbelt though, i guess that isn't a static or kinematic body.
I recently added Velocity and Omega properties to the NewtonBody class.

[Edit]
Added a simple demo to demonstrate the conveyor belt trick. (demo 21)

Regards, Sven-Bertil
